Evdev Remote Control (Bluetooth, BLE, Game Controllers)

drott opened the pull request:
## Pypilot BT Remote

Pypilot remote by grabbing a bluetooth HID/keyboard and sending commands to pypilot.

### Usage

1. Pair a Bluetooth HID device with the machine that is running this script, for example using the OpenPlotter Bluetooth UI, or `bluetoothctl `on the command line, using `scan le`, `pair <MAC>`, then `trust <MAC>`, `connect <MAC>`. where `<MAC>` is the BLE MAC address of the device.
2. Run `# python3 ble_remote.py <BT Mac Address>` and specify a Bluetooth MAC address to use **or** leave empty for menu selection, specify multiple BT Mac addresses for listening for key presses from multiple remotes.

### Requirements

* Python 3.11
* Install py-evdev into your virtual environment, for exampe using `$ pip3 install py-evdev`

### Example

```
$ python3 ble_remote.py 
Available devices:
0: SmartRemote (Path: /dev/input/event6, Unique ID: 4c:5e:a5:86:19:02)
1: SmartRemote (Path: /dev/input/event5, Unique ID: 6b:cf:42:97:cd:74)
2: Jieli Technology UACDemoV1.0 (Path: /dev/input/event4, Unique ID: 4150344C3631350C)
3: vc4-hdmi-1 HDMI Jack (Path: /dev/input/event3, Unique ID: )
4: vc4-hdmi-1 (Path: /dev/input/event2, Unique ID: )
5: vc4-hdmi-0 HDMI Jack (Path: /dev/input/event1, Unique ID: )
6: vc4-hdmi-0 (Path: /dev/input/event0, Unique ID: )
Enter the number of the device you want to select: 0
2025-05-17 20:27:59,645 [INFO] __main__: Listening for events. Press Ctrl+C to stop.
2025-05-17 20:27:59,648 [INFO] pypilot_state: Attempting to connect...
2025-05-17 20:27:59,669 [INFO] pypilot_state: Connected to localhost:23322
2025-05-17 20:27:59,669 [INFO] pypilot_state: Connected successfully.
2025-05-17 20:27:59,670 [INFO] pypilot_state: Starting tasks...
2025-05-17 20:28:00,172 [INFO] pypilot_state: AP State: {'ap.enabled': False, 'ap.heading_command': 84.3, 'ap.heading': 83.5585, 'imu.heading': 83.4889}
2025-05-17 20:28:00,674 [INFO] pypilot_state: AP State: {'ap.enabled': False, 'ap.heading_command': 84.3, 'ap.heading': 84.027, 'imu.heading': 83.664}
2025-05-17 20:28:01,175 [INFO] pypilot_state: AP State: {'ap.enabled': False, 'ap.heading_command': 84.3, 'ap.heading': 84.0482, 'imu.heading': 84.137}
2025-05-17 20:28:01,637 [INFO] __main__: Key: KEY_PLAYPAUSE, State: 1 from SmartRemote
2025-05-17 20:28:01,639 [INFO] __main__: Key: KEY_PLAYPAUSE, State: 0 from SmartRemote
2025-05-17 20:28:01,639 [INFO] pypilot_state: Queued action: Action.TOGGLE
2025-05-17 20:28:01,639 [INFO] pypilot_state: Queued action: Action.ENGAGE
2025-05-17 20:28:01,677 [INFO] pypilot_state: AP State: {'ap.enabled': False, 'ap.heading_command': 83.7, 'ap.heading': 83.7047, 'imu.heading': 83.8166}
2025-05-17 20:28:02,179 [INFO] pypilot_state: AP State: {'ap.enabled': True, 'ap.heading_command': 83.7, 'ap.heading': 83.8554, 'imu.heading': 83.6531}
^C2025-05-17 20:28:02,341 [INFO] __main__: 
Stopped listening for events.
```

sailoog commented:
This looks great! We could implement a GUI for this in the openplotter-pypilot app when it is merged.

seandepagnier commented:
Is this any bluetooth keyboard?   Could you list a specific device in the top of the file?   I would like to test this, and perhaps support bluetooth remotes, though I am concerned somewhat by latency I am not sure how they perform.

Could you put a short message at the top of each file indicating the author, (could just be the github user name), and the license?   Prefer GPLv3 but can be a different one so long as it is compatible.

Can you explain why not to reuse the pypilot client class in client.py?   This can autodetect pypilot even if it is a remote host and simplifies some of the code, but I am wondering if the reason is you wanted to use async?

Finally, this should probably eventually setup as a service like pypilot_hat and others so that it can automatically run in the background at boot.

drott commented:
Refactored this to be more generically named and move into `pypilot/`.

> Is this any bluetooth keyboard? Could you list a specific device in the top of the file?

It is any key-event generating evdev-compatible device. I tested this with an Apple keyboard (non BLE, connection-oriented Bluetooth), "SmartRemote" BLE media buttons from AliExpress, and an Xbox game controller. The latter, connected via USB, but it should work with BT as well.

> I would like to test this, and perhaps support bluetooth remotes, though I am concerned somewhat by latency I am not sure how they perform.

Latency looks good here, events are pretty instance, I don't see a noticeable difference compared to the gamepad when connected over USB.

> Could you put a short message at the top of each file indicating the author, (could just be the github user name), and the license? Prefer GPLv3 but can be a different one so long as it is compatible.

Added license header with Apache 2 license, and author.

> Can you explain why not to reuse the pypilot client class in client.py? This can autodetect pypilot even if it is a remote host and simplifies some of the code, but I am wondering if the reason is you wanted to use async?

I refactored this now to use `client.py`. Indeed, I thought it would be a good idea to use async for evdev event retrieval and for socket events. (A future suggestion and an read that I could imagine contributing to would be to refactor client.py to use async.)

> Finally, this should probably eventually setup as a service like pypilot_hat and others so that it can automatically run in the background at boot.

Agree, I was going to locally run this as a systemd unit. If there is other setup scripts to be touched, etc.: I am not familiar with that and perhaps that could be a separate change. 

Also, two more design thoughts:
1) What do you think about having a JSON string as a datum on the server that defines the keymap? That would make it easier to make a UI as @sailoog suggested to define the keymappings, but I am open to pretty much any other way of doing that.
2) In general, the functionality of action handling here is quite similar to hat.py - and I thought both could actually be on the server side and the server would receive action intents or similar. Or the behaviour parts of hat.py and remote.py could be somehow joined.
